File: skills/reel-studio/scripts/matte-ink.py
"""
Alpha for flat single-colour shapes — ink blots, silhouettes, solid graphics.

Chroma-keying these leaves a magenta halo on every anti-aliased edge pixel,
because the edge is a blend of shape and backdrop. Rebuilding alpha from
darkness and forcing the fill colour cannot fringe at all.

Usage:
  python3 scripts/matte-ink.py ink-splatter.png            # defaults to brand slate
  python3 scripts/matte-ink.py logo-shape.png "#2C6BAC"
"""
import json
import os
import sys

import numpy as np
from PIL import Image

DIR = os.path.join(os.getcwd(), "public", "art")


def brand_slate():
    for candidate in ("src/brand.json", "brand.config.json"):
        try:
            with open(candidate, encoding="utf-8") as fh:
                return json.load(fh)["colors"]["slate"]
        except (OSError, ValueError, KeyError):
            continue
    return "#0F172A"


def to_rgb(hex_colour):
    h = hex_colour.lstrip("#")
    return tuple(int(h[i : i + 2], 16) for i in (0, 2, 4))


name = sys.argv[1] if len(sys.argv) > 1 else "ink-splatter.png"
fill = to_rgb(sys.argv[2] if len(sys.argv) > 2 else brand_slate())

src = os.path.join(DIR, name)
a = np.asarray(Image.open(src).convert("RGB")).astype(np.float32)

# The magenta backdrop is bright; the shape is dark. Alpha is how far below the
# backdrop's luminance each pixel sits.
alpha = np.clip((165.0 - a.mean(axis=2)) / 55.0, 0.0, 1.0)

rgba = np.zeros((*alpha.shape, 4), dtype=np.uint8)
rgba[..., 0], rgba[..., 1], rgba[..., 2] = fill
rgba[..., 3] = (alpha * 255).astype(np.uint8)

img = Image.fromarray(rgba, "RGBA")
img = img.crop(img.getchannel("A").point(lambda v: 255 if v > 10 else 0).getbbox())
img.save(src)

print(f"{name}  {img.width}x{img.height}  {(np.asarray(img)[..., 3] > 10).mean() * 100:.0f}% opaque")

File: skills/reel-studio/scripts/matte-ink.mjs
/**
 * Alpha for flat single-colour shapes — ink blots, silhouettes, solid graphics.
 *
 * Chroma-keying these leaves a magenta halo on every anti-aliased edge pixel,
 * because the edge is a blend of shape and backdrop. Rebuilding alpha from
 * darkness and forcing the fill colour cannot fringe at all.
 *
 * Usage:
 *   node scripts/matte-ink.mjs ink-splatter.png            # defaults to brand slate
 *   node scripts/matte-ink.mjs logo-shape.png "#2C6BAC"
 */
import fs from "fs";
import path from "path";
import { Jimp } from "jimp";

const DIR = path.join(process.cwd(), "public", "art");

const brandSlate = () => {
  for (const candidate of ["src/brand.json", "brand.config.json"]) {
    try {
      return JSON.parse(fs.readFileSync(candidate, "utf8")).colors.slate;
    } catch {
      /* try the next one */
    }
  }
  return "#0F172A";
};

const toRgb = (hex) => {
  const h = hex.replace("#", "");
  return [0, 2, 4].map((i) => parseInt(h.slice(i, i + 2), 16));
};

const name = process.argv[2] ?? "ink-splatter.png";
const [fr, fg, fb] = toRgb(process.argv[3] ?? brandSlate());

const img = await Jimp.read(path.join(DIR, name));
const { width, height, data } = img.bitmap;

let minX = width;
let minY = height;
let maxX = -1;
let maxY = -1;

for (let i = 0; i < data.length; i += 4) {
  // The magenta backdrop is bright; the shape is dark. Alpha is how far below
  // the backdrop's luminance each pixel sits.
  const lum = (data[i] + data[i + 1] + data[i + 2]) / 3;
  const alpha = Math.round(255 * Math.min(1, Math.max(0, (165 - lum) / 55)));

  data[i] = fr;
  data[i + 1] = fg;
  data[i + 2] = fb;
  data[i + 3] = alpha;

  if (alpha > 10) {
    const px = (i / 4) % width;
    const py = Math.floor(i / 4 / width);
    if (px < minX) minX = px;
    if (px > maxX) maxX = px;
    if (py < minY) minY = py;
    if (py > maxY) maxY = py;
  }
}

if (maxX < 0) throw new Error(`${name}: nothing dark enough to keep`);

img.crop({ x: minX, y: minY, w: maxX + 1 - minX, h: maxY + 1 - minY });
await img.write(path.join(DIR, name));

const out = img.bitmap;
let opaque = 0;
for (let i = 3; i < out.data.length; i += 4) if (out.data[i] > 10) opaque++;
console.log(
  `${name}  ${out.width}x${out.height}  ${((opaque / (out.width * out.height)) * 100).toFixed(0)}% opaque`,
);
